moveit_ros/moveit_servo/include/moveit_servo/smoothing_plugin/low_pass_filter.h

```diff
@@ -0,0 +1,118 @@
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Andy Zelenak
+   Description: A first-order Butterworth low-pass filter. There is only one parameter to tune.
+ */
+
+#pragma once
+
+#include <cstddef>
+
+#include <moveit/robot_model/robot_model.h>
+#include <moveit_servo/servo_parameters.h>
+#include <moveit_servo/smoothing_plugin/smoothing_base_class.h>
+
+namespace moveit_servo
+{
+/**
+ * Class LowpassFilterImpl - Implementation of a signal filter to soften jerks.
+ * This is a first-order Butterworth low-pass filter. First-order was chosen for 2 reasons:
+ * - It doesn't overshoot
+ * - Computational efficiency
+ */
+class LowpassFilterImpl
+{
+public:
+  /**
+   * Constructor.
+   * @param low_pass_filter_coeff Larger filter_coeff-> more smoothing of servo commands, but more lag.
+   * Rough plot, with cutoff frequency on the y-axis:
+   * https://www.wolframalpha.com/input/?i=plot+arccot(c)
+   */
+  LowpassFilterImpl(double low_pass_filter_coeff);
+
+  double filter(double new_measurement);
+
+  void reset(double data);
+
+private:
+  static constexpr std::size_t FILTER_LENGTH = 2;
+  double previous_measurements_[FILTER_LENGTH];
+  double previous_filtered_measurement_;
+  // Scale and feedback term are calculated from supplied filter coefficient
+  double scale_term_;
+  double feedback_term_;
+};
+
+// Plugin
+class LowPassFilter : public SmoothingBaseClass
+{
+public:
+  LowPassFilter(){};
+
+  /**
+   * Initialize the smoothing algorithm
+   * @param node ROS node, used for parameter retrieval
+   * @param robot_model typically used to retrieve vel/accel/jerk limits
+   * @param num_joints number of actuated joints in the JointGroup Servo controls
+   * @param parameters access to all Servo parameters that have been parsed from yaml
+   * @return True if initialization was successful
+   */
+  bool initialize(rclcpp::Node::SharedPtr node, moveit::core::RobotModelConstPtr robot_model, const size_t num_joints,
```

Comment:
Avoid `const` value types in declaration (but keep them in definition) https://abseil.io/tips/109

Reply:
Meh, I'm not sure I agree with that. The `const` in the header file is part of documenting usage and it should match what's in the cpp file.

Let me know if I misinterpreted your comment.

Reply:
The point of the tip is that C++ cannot disambiguate between `const` value and non-`const` value passing so `void f(int x)` and `void f(const int x)` are the same. As abseil is not part of our style guide pipeline, it's your option. As far as usage, it's a copy so it doesn't affect the caller.

Reply:
I did read the abseil thing. Basically, I think it says:

void F(const int);               //    <--  use const here
void F(const int) { /* ... */ }  // <-- do not use const here

Which is different from what you're saying. I agree with you though, so I'll change it

Reply:
The bottom of the page summarizes their guidance
> 1. Never use top-level const on function parameters in declarations that are not definitions (and be careful not to copy/paste a meaningless const). It is meaningless and ignored by the compiler, it is visual noise, and it could mislead readers.
> 2. Do use top-level const on function parameters in definitions at your (or your team’s) discretion. You might follow the same rationale as you would for when to declare a function-local variable const.

```
void F(const int); // Declaration
void F(const int) { /* ... */ } // Definition
```